waiter/src/waiter/state.clj

```diff
@@ -1303,3 +1308,114 @@
     (t/millis router-syncer-interval-ms)
     #(retrieve-peer-routers discovery router-chan)
     :delay-ms router-syncer-delay-ms))
+
+;; Support for tracking instance scheduling/startup time stats
+
+(defrecord InstanceTracker [service-id known-instance-ids scheduling-instance-timer-contexts starting-instance-trackers])
+
+(defn- new-tracker-val
+  [service-id service-id->service-description-fn]
+  (let [{:strs [min-instances]} (service-id->service-description-fn service-id)
+        ctxs (repeatedly
+               min-instances
+               (fn [] {:schedule [(timers/start (metrics/waiter-timer "service-init-overhead" "schedule-time"))
+                                  (timers/start (metrics/service-timer service-id "service-init-overhead" "schedule-time"))]
+                       :total [(timers/start (metrics/waiter-timer "service-init-overhead" "init-to-healthy-time"))
+                               (timers/start (metrics/service-timer service-id "service-init-overhead" "init-to-healthy-time"))]}))]
+    (map->InstanceTracker
+      {:service-id service-id
+       :known-instance-ids #{}
+       :scheduling-instance-timer-contexts (vec ctxs)
+       :starting-instance-trackers []})))
+
+(defn update-instance-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-instance-trackers new-service-ids removed-service-ids service-id->healthy-instances service-id->unhealthy-instances service-id->service-description-fn]
+  (->> service-instance-trackers
+       (remove (comp removed-service-ids :service-id))
+       (concat (mapv #(new-tracker-val % service-id->service-description-fn) new-service-ids))
+       (mapv (fn [{:keys [service-id known-instance-ids scheduling-instance-timer-contexts starting-instance-trackers]}]
+               (let [healthy-instance-ids (->> service-id
+                                               (get service-id->healthy-instances)
+                                               (map :id)
+                                               set)
+                     known-instance-ids' (->> service-id
+                                              (get service-id->unhealthy-instances)
+                                              (map :id)
+                                              (into healthy-instance-ids))
```

Comment:
Why does something that becomes temporarily unhealthy become an unknown instance id? 

Is a one-shot attempt at pairing timers and instance no longer possible if an instance state can change between healthy and unhealthy? If so, I'd prefer starting out with a bank of unpaired times, and then pairing them up whenever we can. Presence of a previously unknown unhealthy instance is also a good time to associate the instance with new timer contexts.

Reply:
I think you misread this. `known-instance-ids` is the union of all healthy and unhealthy instances.

Reply:
Oops, my bad.